File: packages/backstage-go/cron.go
package backstage

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

type CronTask struct {
	Schedule string
	TaskName string
	Queue    *Queue
	Args     []interface{}
	fields   cronFields
	lastRun  time.Time
}

type cronFields struct {
	minute     []int
	hour       []int
	dayOfMonth []int
	month      []int
	dayOfWeek  []int
}

func NewCronTask(schedule, taskName string, queue *Queue, args ...interface{}) (*CronTask, error) {
	ct := &CronTask{
		Schedule: schedule,
		TaskName: taskName,
		Queue:    queue,
		Args:     args,
	}

	fields, err := parseCron(schedule)
	if err != nil {
		return nil, err
	}
	ct.fields = fields

	return ct, nil
}

func parseCron(schedule string) (cronFields, error) {
	parts := strings.Fields(schedule)
	if len(parts) != 5 {
		return cronFields{}, fmt.Errorf("invalid cron: expected 5 fields, got %d", len(parts))
	}

	minute, err := parseField(parts[0], 0, 59)
	if err != nil {
		return cronFields{}, err
	}

	hour, err := parseField(parts[1], 0, 23)
	if err != nil {
		return cronFields{}, err
	}

	dayOfMonth, err := parseField(parts[2], 1, 31)
	if err != nil {
		return cronFields{}, err
	}

	month, err := parseField(parts[3], 1, 12)
	if err != nil {
		return cronFields{}, err
	}

	dayOfWeek, err := parseField(parts[4], 0, 6)
	if err != nil {
		return cronFields{}, err
	}

	return cronFields{
		minute:     minute,
		hour:       hour,
		dayOfMonth: dayOfMonth,
		month:      month,
		dayOfWeek:  dayOfWeek,
	}, nil
}

func parseField(field string, min, max int) ([]int, error) {
	values := make(map[int]bool)

	for _, part := range strings.Split(field, ",") {
		if part == "*" {
			for i := min; i <= max; i++ {
				values[i] = true
			}
		} else if strings.Contains(part, "/") {
			split := strings.Split(part, "/")
			step, err := strconv.Atoi(split[1])
			if err != nil || step <= 0 {
				return nil, fmt.Errorf("invalid step: %s", part)
			}

			start, end := min, max
			if split[0] != "*" {
				if strings.Contains(split[0], "-") {
					rangeParts := strings.Split(split[0], "-")
					start, _ = strconv.Atoi(rangeParts[0])
					end, _ = strconv.Atoi(rangeParts[1])
				} else {
					start, _ = strconv.Atoi(split[0])
				}
			}

			for i := start; i <= end; i += step {
				values[i] = true
			}
		} else if strings.Contains(part, "-") {
			rangeParts := strings.Split(part, "-")
			start, _ := strconv.Atoi(rangeParts[0])
			end, _ := strconv.Atoi(rangeParts[1])
			for i := start; i <= end; i++ {
				values[i] = true
			}
		} else {
			val, err := strconv.Atoi(part)
			if err != nil {
				return nil, fmt.Errorf("invalid value: %s", part)
			}
			values[val] = true
		}
	}

	result := make([]int, 0, len(values))
	for v := range values {
		if v < min || v > max {
			return nil, fmt.Errorf("value %d out of range [%d-%d]", v, min, max)
		}
		result = append(result, v)
	}
	return result, nil
}

func (ct *CronTask) NextRun(after time.Time) time.Time {
	next := after.Truncate(time.Minute).Add(time.Minute)

	for i := 0; i < 525600; i++ { // max 1 year
		if ct.matches(next) {
			return next
		}
		next = next.Add(time.Minute)
	}

	return time.Time{}
}

func (ct *CronTask) matches(t time.Time) bool {
	return contains(ct.fields.minute, t.Minute()) &&
		contains(ct.fields.hour, t.Hour()) &&
		contains(ct.fields.dayOfMonth, t.Day()) &&
		contains(ct.fields.month, int(t.Month())) &&
		contains(ct.fields.dayOfWeek, int(t.Weekday()))
}

func contains(slice []int, val int) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func (ct *CronTask) MarkRun(at time.Time) {
	ct.lastRun = at
}

File: packages/backstage-ts/src/cron.ts
/**
 * Backstage cron task definitions.
 */

import { InvalidCronSchedule } from './exceptions';
import type { Queue } from './queue';

interface CronFields {
  minute: number[];
  hour: number[];
  dayOfMonth: number[];
  month: number[];
  dayOfWeek: number[];
}

export class CronTask {
  public readonly schedule: string;
  public readonly taskName: string;
  public readonly queue?: Queue;
  public readonly args: unknown[];

  private fields: CronFields;
  private lastRun?: Date;

  constructor(
    schedule: string,
    taskName: string,
    queue?: Queue,
    args: unknown[] = []
  ) {
    this.schedule = schedule;
    this.taskName = taskName;
    this.queue = queue;
    this.args = args;
    this.fields = this.parseCron(schedule);
  }

  private parseCron(schedule: string): CronFields {
    const parts = schedule.trim().split(/\s+/);

    if (parts.length !== 5) {
      throw new InvalidCronSchedule(
        schedule,
        'Expected 5 fields: minute hour dayOfMonth month dayOfWeek'
      );
    }

    return {
      minute: this.parseField(parts[0]!, 0, 59),
      hour: this.parseField(parts[1]!, 0, 23),
      dayOfMonth: this.parseField(parts[2]!, 1, 31),
      month: this.parseField(parts[3]!, 1, 12),
      dayOfWeek: this.parseField(parts[4]!, 0, 6),
    };
  }

  private parseField(field: string, min: number, max: number): number[] {
    const values: Set<number> = new Set();
    const parts = field.split(',');

    for (const part of parts) {
      if (part === '*') {
        for (let i = min; i <= max; i++) {
          values.add(i);
        }
      } else if (part.includes('/')) {
        const [range, stepStr] = part.split('/');
        const step = parseInt(stepStr ?? '1', 10);
        if (isNaN(step) || step <= 0) {
          throw new InvalidCronSchedule(this.schedule, `Invalid step: ${part}`);
        }

        let start = min;
        let end = max;

        if (range && range !== '*') {
          if (range.includes('-')) {
            const [s, e] = range.split('-');
            start = parseInt(s ?? String(min), 10);
            end = parseInt(e ?? String(max), 10);
          } else {
            start = parseInt(range, 10);
          }
        }

        for (let i = start; i <= end; i += step) {
          values.add(i);
        }
      } else if (part.includes('-')) {
        const [startStr, endStr] = part.split('-');
        const start = parseInt(startStr ?? String(min), 10);
        const end = parseInt(endStr ?? String(max), 10);

        if (isNaN(start) || isNaN(end)) {
          throw new InvalidCronSchedule(
            this.schedule,
            `Invalid range: ${part}`
          );
        }

        for (let i = start; i <= end; i++) {
          values.add(i);
        }
      } else {
        const value = parseInt(part, 10);
        if (isNaN(value)) {
          throw new InvalidCronSchedule(
            this.schedule,
            `Invalid value: ${part}`
          );
        }
        values.add(value);
      }
    }

    return Array.from(values).sort((a, b) => a - b);
  }

  getNextRun(after: Date = new Date()): Date {
    const next = new Date(after.getTime());
    next.setSeconds(0, 0);
    next.setMinutes(next.getMinutes() + 1);

    const maxIterations = 525600;
    let iterations = 0;

    while (iterations < maxIterations) {
      const minute = next.getMinutes();
      const hour = next.getHours();
      const dayOfMonth = next.getDate();
      const month = next.getMonth() + 1;
      const dayOfWeek = next.getDay();

      if (
        this.fields.minute.includes(minute) &&
        this.fields.hour.includes(hour) &&
        this.fields.dayOfMonth.includes(dayOfMonth) &&
        this.fields.month.includes(month) &&
        this.fields.dayOfWeek.includes(dayOfWeek)
      ) {
        return next;
      }

      next.setMinutes(next.getMinutes() + 1);
      iterations++;
    }

    throw new InvalidCronSchedule(
      this.schedule,
      'Could not find next run within a year'
    );
  }

  shouldRun(now: Date = new Date()): boolean {
    if (!this.lastRun) return true;
    const next = this.getNextRun(this.lastRun);
    return now >= next;
  }

  markRun(at: Date = new Date()): void {
    this.lastRun = at;
  }

  reset(to?: Date): void {
    this.lastRun = to;
  }
}
